In [1]:
import numpy as np

arr = np.array([19, 242, 14, 152, 142, 1000])

arr.mean()

261.5

# Введение в Pandas

Данная библиотека построена поверх уже знакомой вам библиотеки NumPy, которая в свою очередь написана на языке С. Такая особенность даёт большой скачок в производительности операций библиотеки.

Ещё одним большим плюсом Pandas является подробная документация с детальным описанием всех функций и примерами их использования. Более того, так как Pandas является очень популярной библиотекой, то вы практически всегда сможете найти ответ на интересующий вас вопрос в интернете, главное — уметь правильно искать

ссылка на документацию https://pandas.pydata.org/pandas-docs/stable/index.html


In [2]:
import pandas as pd

pd.__version__

'1.4.4'

In [3]:
print(pd.__name__)

pandas


# Pandas.Series

Series — это упорядоченная изменяемая коллекция объектов, имеющая так называемые ассоциативные метки (индексы). 

## Создание Series
### Способ 1. Из списка

In [4]:
countries = pd.Series(
    data = ['Англия', 'Канада', 'США', 'Россия', 'Украина', 'Беларусь', 'Казахстан'],
    index = ['UK', 'CA', 'US', 'RU', 'UA', 'BY', 'KZ'],
    name = 'countries'
)
display(countries)

UK       Англия
CA       Канада
US          США
RU       Россия
UA      Украина
BY     Беларусь
KZ    Казахстан
Name: countries, dtype: object

In [5]:
# если не указать index, то он будет присвоен автоматически в виде чисел

countries = pd.Series(
    ['Англия', 'Канада', 'США', 'Россия', 'Украина', 'Беларусь', 'Казахстан']
)
display(countries)

0       Англия
1       Канада
2          США
3       Россия
4      Украина
5     Беларусь
6    Казахстан
dtype: object

### Способ 2. Из словаря

In [6]:
countries = pd.Series({
    'UK': 'Англия',
    'CA': 'Канада',
    'US' : 'США',
    'RU': 'Россия',
    'UA': 'Украина',
    'BY': 'Беларусь',
    'KZ': 'Казахстан'},
    name = 'countries'
)
display(countries)

UK       Англия
CA       Канада
US          США
RU       Россия
UA      Украина
BY     Беларусь
KZ    Казахстан
Name: countries, dtype: object

## Доступ к данным в Series
Доступ к элементам осуществляется с использованием loc или iloc.
На самом деле loc и iloc можно опустить и обращаться к элементам Series напрямую по индексам, например countries[‘UK’, 'US', ‘UA’] или countries[0, 2, 4]. Оба варианта являются равноправными для Series, однако в дальнейшем мы будем использовать эти операции при обращении к более сложной структуре — DataFrame, а в контексте этой структуры эти варианты уже неравноправны.

In [7]:
# loc. можно передать значение метки в квадратных скобках
print(countries.loc['US'])

США


In [8]:
# можно передать несколько значений меток 
print(countries.loc[['US', 'RU', 'UK']])


US       США
RU    Россия
UK    Англия
Name: countries, dtype: object


In [9]:
# iloc - запрос по порядковому индексу
print(countries.iloc[6])
# Казахстан

Казахстан


In [10]:
# также, можно передать несколько значений или срезы. в этом случае возвращается объект Series.
# в данном примере конечное значение диапазона не включается в результат
print(countries.iloc[1:4])

CA    Канада
US       США
RU    Россия
Name: countries, dtype: object


# Pandas.DataFrame
## DataFrame как структура данных
DataFrame является двумерной структурой и представляется в виде таблицы, в которой есть строки и столбцы: столбцами в DataFrame выступают объекты Series, а строки формируются из их элементов. Также в DataFrame есть метки (индексы), которые соответствуют каждой строке таблицы.

## Создание DataFrame
DataFrame создаётся с помощью функции pd.DataFrame(). Так же, как и для Series, для создания объектов DataFrame есть несколько способов:

### Способ 1
Самый простой способ создания DataFrame — из словаря, ключами которого являются имена столбцов будущей таблицы, а значениями — списки, в которых хранится содержимое этих столбцов:

In [11]:
countries_df = pd.DataFrame({
    'country': ['Англия', 'Канада', 'США', 'Россия', 'Украина', 'Беларусь', 'Казахстан'],
    'population': [56.29, 38.05, 322.28, 146.24, 45.5, 9.5, 17.04],
    'square': [133396, 9984670, 9826630, 17125191, 603628, 207600, 2724902]
})

display(countries_df)

country  population    square
0     Англия       56.29    133396
1     Канада       38.05   9984670
2        США      322.28   9826630
3     Россия      146.24  17125191
4    Украина       45.50    603628
5   Беларусь        9.50    207600
6  Казахстан       17.04   2724902

In [12]:
# Обратите внимание, что, так как мы не задали метки (индексы) DataFrame, они были сгенерированы автоматически. Исправим это, задав индексы вручную:
countries_df.index = ['UK', 'CA', 'US', 'RU', 'UA', 'BY', 'KZ']
display(countries_df)

country  population    square
UK     Англия       56.29    133396
CA     Канада       38.05   9984670
US        США      322.28   9826630
RU     Россия      146.24  17125191
UA    Украина       45.50    603628
BY   Беларусь        9.50    207600
KZ  Казахстан       17.04   2724902

### Способ 2.
Также DataFrame можно создать из вложенного списка, внутренние списки которого будут являться строками новой таблицы:

In [13]:
countries_df = pd.DataFrame(
    data = [
        ['Англия', 56.29, 133396],
        ['Канада', 38.05, 9984670],
        ['США', 322.28, 9826630],
        ['Россия', 146.24, 17125191],
        ['Украина', 45.5, 603628],
        ['Беларусь', 9.5, 207600],
        ['Казахстан', 17.04, 2724902]
    ],
    columns= ['country', 'population', 'square'],
    index = ['UK', 'CA', 'US', 'RU', 'UA', 'BY', 'KZ']
)
display(countries_df)

country  population    square
UK     Англия       56.29    133396
CA     Канада       38.05   9984670
US        США      322.28   9826630
RU     Россия      146.24  17125191
UA    Украина       45.50    603628
BY   Беларусь        9.50    207600
KZ  Казахстан       17.04   2724902

## AXIS в DataFrame
При работе с Pandas важно уметь указывать направление работы метода, который используется. Для этого вводится понятие axis (ось, координата). Движение по строкам в таблице обозначается axis с индексом 0, а движение по столбцам — axis с индексом 1.

Данный параметр заложен во все методы, которые могут работать в двух направлениях и по умолчанию в большинстве из них axis=0, то есть они выполняют операции со строками, если не задавать axis вручную.

In [14]:
# Считаем среднее по строкам (axis = 0) в каждом столбце:
countries_df.mean(axis=0, numeric_only=True)

population    9.070000e+01
square        5.800860e+06
dtype: float64

In [15]:
#Считаем среднее по столбцам (axis = 1) в каждой строке:
countries_df.mean(axis=1, numeric_only=True)

UK      66726.145
CA    4992354.025
US    4913476.140
RU    8562668.620
UA     301836.750
BY     103804.750
KZ    1362459.520
dtype: float64

## Доступ к данным в DataFrame

In [16]:
# Можно обратиться к DataFrame по имени столбца через точку:
countries_df.population

# Однако использование такого способа возможно только тогда, когда имя столбца указано без пробелов.

UK     56.29
CA     38.05
US    322.28
RU    146.24
UA     45.50
BY      9.50
KZ     17.04
Name: population, dtype: float64

In [17]:
# Другой вариант — обратиться к DataFrame по индексу и указать имя столбца:
countries_df['population']

UK     56.29
CA     38.05
US    322.28
RU    146.24
UA     45.50
BY      9.50
KZ     17.04
Name: population, dtype: float64

In [18]:
# при этом мы получаем объект Series

type(countries_df.population)

pandas.core.series.Series

 Для того чтобы получить доступ к ячейкам таблицы, используются уже знакомые нам loc и iloc.
 При этом, в соответствии с механизмом работы axis, при обращении к DataFrame по индексам с помощью loc (iloc) первым индексом указывается индекс (порядковый номер), соответствующий строкам, а вторым — имя (порядковый номер) столбца.


In [19]:
# Получим площадь Великобритании:
countries_df.loc['UK', 'square']

133396

In [20]:
# Получим население и площадь, соответствующие России:
countries_df.loc['RU', ['population', 'square']]

population      146.24
square        17125191
Name: RU, dtype: object

In [21]:
# Сделаем вырезку из таблицы и получим информацию о населении и площади, соответствующую Украине, Беларуси и Казахстану:
countries_df.loc[['UA', 'BY', 'KZ'],['population', 'square']]

population   square
UA       45.50   603628
BY        9.50   207600
KZ       17.04  2724902

In [22]:
# или
countries_df.iloc[4:7, 1:3]

population   square
UA       45.50   603628
BY        9.50   207600
KZ       17.04  2724902

#  Работа с различными источниками данных в Pandas

## Запись в CSV файл

Предположим, что мы захотели сохранить созданный нами ранее DataFrame. Самым простым и распространённым источником табличных данных является формат csv (comma-separated values). В данном формате ячейки таблицы обозначаются некоторым разделителем, чаще всего запятой либо точкой с запятой.

Экспорт данных в формат csv осуществляется с помощью метода DataFrame to_csv().  
path_or_buf — путь до файла, в который будет записан DataFrame (например, data/my_data.csv);
sep — разделитель данных в выходном файле (по умолчанию ',');
decimal — разделитель чисел на целую и дробную части в выходном файле (по умолчанию '.');
columns — список столбцов, которые нужно записать в файл (по умолчанию записываются все столбцы);
index — параметр, определяющий, требуется ли создавать дополнительный столбец с индексами строк в файле (по умолчанию True).

In [23]:
# Заранее создадим папку data в директории, где лежит наш ноутбук. Теперь давайте сохраним наш DataFrame с информацией о странах 
# в csv-файл countries.csv и положим файл в папку data. При этом укажем, что разделителем в нашем файле будет являться символ ';', 
# а также то, что нам не нужен дополнительный столбец с индексами строк:

countries_df = pd.DataFrame({
    'country': ['Англия', 'Канада', 'США', 'Россия', 'Украина', 'Беларусь', 'Казахстан'],
    'population': [56.29, 38.05, 322.28, 146.24, 45.5, 9.5, 17.04],
    'square': [133396, 9984670, 9826630, 17125191, 603628, 207600, 2724902]
})

countries_df.to_csv('data/countries.csv', index=False, sep=';')

## Чтение из CSV файла
Для чтения таблицы из csv-файла используется функция модуля Pandas read_csv. Функция возвращает DataFrame и имеет несколько важных параметров.
filepath_or_buffer — путь до файла, который мы читаем;
sep — разделитель данных (по умолчанию ',');
decimal — разделитель чисел на целую и дробную часть в выходном файле (по умолчанию '.');
names — список с названиями столбцов для чтения;
skiprows — количество строк в файле, которые нужно пропустить (например, файл может содержать служебную информацию, которая нам не нужна).

In [24]:
# Убедимся, что сохранённый нами ранее файл создался верно. Для этого прочитаем его в переменную countries_data и 
# выведем её на экран. Не забудем также о том, что мы использовали в качестве разделителя ';':

countries_data = pd.read_csv('data/countries.csv', sep=';')
display(countries_data)

country  population    square
0     Англия       56.29    133396
1     Канада       38.05   9984670
2        США      322.28   9826630
3     Россия      146.24  17125191
4    Украина       45.50    603628
5   Беларусь        9.50    207600
6  Казахстан       17.04   2724902

## Чтение CSV файла по ссылке

На самом деле файл с данными не обязательно должен храниться у вас на компьютере. Если он находится в открытом доступе по ссылке (например, на Google Диске или GitHub), его можно прочитать и из интернета — для этого достаточно в функции read_csv() вместо пути до файла указать ссылку на файл. Например:

In [25]:
data = pd.read_csv('https://raw.githubusercontent.com/esabunor/MLWorkspace/master/melb_data.csv')
display(data)

Unnamed: 0        Suburb           Address  Rooms Type      Price  \
0               1    Abbotsford      85 Turner St      2    h  1480000.0   
1               2    Abbotsford   25 Bloomburg St      2    h  1035000.0   
2               4    Abbotsford      5 Charles St      3    h  1465000.0   
3               5    Abbotsford  40 Federation La      3    h   850000.0   
4               6    Abbotsford       55a Park St      4    h  1600000.0   
...           ...           ...               ...    ...  ...        ...   
18391       23540  Williamstown   8/2 Thompson St      2    t   622500.0   
18392       23541  Williamstown      96 Verdon St      4    h  2500000.0   
18393       23544     Yallambie      17 Amaroo Wy      4    h  1100000.0   
18394       23545    Yarraville        6 Agnes St      4    h  1285000.0   
18395       23546    Yarraville     33 Freeman St      4    h  1050000.0   

      Method     SellerG        Date  Distance  ...  Bathroom  Car  Landsize  \
0          S      Biggin   3/12/2016       2.5  ...       1.0  1.0     202.0   
1          S      Biggin   4/02/2016       2.5  ...       1.0  0.0     156.0   
2         SP      Biggin   4/03/2017       2.5  ...       2.0  0.0     134.0   
3         PI      Biggin   4/03/2017       2.5  ...       2.0  1.0      94.0   
4         VB      Nelson   4/06/2016       2.5  ...       1.0  2.0     120.0   
...      ...         ...         ...       ...  ...       ...  ...       ...   
18391     SP        Greg  26/08/2017       6.8  ...       2.0  1.0       NaN   
18392     PI     Sweeney  26/08/2017       6.8  ...       1.0  5.0     866.0   
18393      S  Buckingham  26/08/2017      12.7  ...       3.0  2.0       NaN   
18394     SP     Village  26/08/2017       6.3  ...       1.0  1.0     362.0   
18395     VB     Village  26/08/2017       6.3  ...       2.0  2.0       NaN   

       BuildingArea  YearBuilt  CouncilArea  Lattitude Longtitude  \
0               NaN        NaN        Yarra  -37.79960  144.99840   
1              79.0     1900.0        Yarra  -37.80790  144.99340   
2             150.0     1900.0        Yarra  -37.80930  144.99440   
3               NaN        NaN        Yarra  -37.79690  144.99690   
4             142.0     2014.0        Yarra  -37.80720  144.99410   
...             ...        ...          ...        ...        ...   
18391          89.0     2010.0          NaN  -37.86393  144.90484   
18392         157.0     1920.0          NaN  -37.85908  144.89299   
18393           NaN        NaN          NaN  -37.72006  145.10547   
18394         112.0     1920.0          NaN  -37.81188  144.88449   
18395         139.0     1950.0          NaN  -37.81829  144.87404   

                  Regionname  Propertycount  
0      Northern Metropolitan         4019.0  
1      Northern Metropolitan         4019.0  
2      Northern Metropolitan         4019.0  
3      Northern Metropolitan         4019.0  
4      Northern Metropolitan         4019.0  
...                      ...            ...  
18391   Western Metropolitan         6380.0  
18392   Western Metropolitan         6380.0  
18393  Northern Metropolitan         1369.0  
18394   Western Metropolitan         6543.0  
18395   Western Metropolitan         6543.0  

[18396 rows x 22 columns]

## Запись и чтение в других форматах
Как уже говорилось ранее, Pandas способен работать со многими распространёнными форматами данных.

Методы для записи таблиц в файлы отличных от csv форматов:

to_excel() — запись DataFrame в формат Excel-таблицы (.xslx);
to_json() — запись DataFrame в формат JSON (.json);
to_xml() — запись DataFrame в формат XML-документа (.xml);
to_sql() — запись DataFrame в базу данных SQL (для реализации этого метода необходимо установить соединение с базой данных).

Методы для чтения таблиц из файлов в отличных от csv форматах:

read_excel() — чтение из формата Excel-таблицы(.xslx) в DataFrame;
read_json() — чтение из формата JSON (.json) в DataFrame;
read_xml() — чтение из формата XML-документа (.xml) в DataFrame;
read_sql() — чтение из базы данных SQL в DataFrame (также необходимо установить соединение с базой данных).

# Знакомимся с денными. Недвижимость

Данные в data/melb_data.csv представляют собой таблицу, в которой содержится 23 столбца:

* index — номер строки
* Suburb — наименование пригорода
* Address — адрес
* Rooms — количество комнат в помещении
* Type — тип здания (h — дом, коттедж, вилла, терраса; u — блочный, дуплексный дом; t — таунхаус)
* Price — цена помещения
* Method — метод продажи 
* SellerG — риэлторская компания
* Date — дата продажи (в формате день/месяц/год)
* Distance — расстояния до объекта от центра Мельбурна 
* Postcode — почтовый индекс
* Bedroom — количество спален
* Bathroom — количество ванных комнат
* Car — количество парковочных мест
* Landsize — площадь прилегающей территории
* BuildingArea — площадь здания
* YearBuilt — год постройки
* CouncilArea — региональное управление
* Lattitude — географическая широта
* Longitude — географическая долгота
* Regionname — наименование района Мельбурна
* Propertycount — количество объектов недвижимости в районе
* Coordinates — широта и долгота, объединённые в кортеж*

In [26]:
# Прочитаем наши данные о недвижимости из csv-файла и запишем результирующий DataFrame в переменную melb_data:
melb_data = pd.read_csv('data/melb_data.csv', sep=',')

In [27]:
# Какова цена объекта недвижимости под индексом 15?

melb_data.loc[15, 'Price']

1310000.0

In [28]:
# Когда был продан объект под индексом 90?
melb_data.loc[90, 'Date']

'10/09/2016'

In [29]:
# Во сколько раз площадь прилегающей территории, на которой находится здание с индексом 3521, больше площади участка,
#  на котором находится здание с индексом 1690? Ответ округлите до целого числа.

square_more = round(melb_data.loc[3521, 'Landsize'] / melb_data.loc[1690, 'Landsize'], 0)
display(square_more)

melb_data.loc[3521, 'Landsize']
melb_data.loc[1690, 'BuildingArea']

3.0

148.0

# Исследование структуры DataFrame
## head() и tail()
Часто бывает такое, что вывести на экран все строки таблицы является ресурсозатратной операцией, а иногда и вовсе не представляется возможным. В большинстве случаев для того, чтобы понять структуру DataFrame и удостовериться, что таблица подгрузилась верно, достаточно вывести несколько первых или последних строк.
Для этого у DataFrame есть методы head() и tail(), которые возвращают n первых и n последних строк таблицы соответственно (по умолчанию n = 5).

In [30]:
# Выведем первые пять строк нашей таблицы:
display(melb_data.head())

index      Suburb           Address  Rooms Type      Price Method SellerG  \
0      0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1      1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2      2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3      3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4      4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  ...  Car  Landsize  BuildingArea  YearBuilt  \
0  3/12/2016       2.5  ...  1.0     202.0         126.0     1970.0   
1  4/02/2016       2.5  ...  0.0     156.0          79.0     1900.0   
2  4/03/2017       2.5  ...  0.0     134.0         150.0     1900.0   
3  4/03/2017       2.5  ...  1.0      94.0         126.0     1970.0   
4  4/06/2016       2.5  ...  2.0     120.0         142.0     2014.0   

   CouncilArea  Lattitude  Longtitude             Regionname  Propertycount  \
0        Yarra   -37.7996    144.9984  Northern Metropolitan         4019.0   
1        Yarra   -37.8079    144.9934  Northern Metropolitan         4019.0   
2        Yarra   -37.8093    144.9944  Northern Metropolitan         4019.0   
3        Yarra   -37.7969    144.9969  Northern Metropolitan         4019.0   
4        Yarra   -37.8072    144.9941  Northern Metropolitan         4019.0   

          Coordinates  
0  -37.7996, 144.9984  
1  -37.8079, 144.9934  
2  -37.8093, 144.9944  
3  -37.7969, 144.9969  
4  -37.8072, 144.9941  

[5 rows x 23 columns]

In [31]:
# Следующий код выведет семь последних строк нашей таблицы:
melb_data.tail(7)

index         Suburb        Address  Rooms Type      Price Method  \
13573  13573       Werribee    5 Nuragi Ct      4    h   635000.0      S   
13574  13574    Westmeadows     9 Black St      3    h   582000.0      S   
13575  13575  Wheelers Hill   12 Strada Cr      4    h  1245000.0      S   
13576  13576   Williamstown  77 Merrett Dr      3    h  1031000.0     SP   
13577  13577   Williamstown    83 Power St      3    h  1170000.0      S   
13578  13578   Williamstown   96 Verdon St      4    h  2500000.0     PI   
13579  13579     Yarraville     6 Agnes St      4    h  1285000.0     SP   

             SellerG        Date  Distance  ...  Car  Landsize  BuildingArea  \
13573  hockingstuart  26/08/2017      14.7  ...  1.0     662.0         172.0   
13574            Red  26/08/2017      16.5  ...  2.0     256.0         126.0   
13575          Barry  26/08/2017      16.7  ...  2.0     652.0         126.0   
13576       Williams  26/08/2017       6.8  ...  2.0     333.0         133.0   
13577          Raine  26/08/2017       6.8  ...  4.0     436.0         126.0   
13578        Sweeney  26/08/2017       6.8  ...  5.0     866.0         157.0   
13579        Village  26/08/2017       6.3  ...  1.0     362.0         112.0   

       YearBuilt  CouncilArea  Lattitude  Longtitude  \
13573     1980.0          NaN  -37.89327   144.64789   
13574     1970.0          NaN  -37.67917   144.89390   
13575     1981.0          NaN  -37.90562   145.16761   
13576     1995.0          NaN  -37.85927   144.87904   
13577     1997.0          NaN  -37.85274   144.88738   
13578     1920.0          NaN  -37.85908   144.89299   
13579     1920.0          NaN  -37.81188   144.88449   

                       Regionname  Propertycount           Coordinates  
13573        Western Metropolitan        16166.0  -37.89327, 144.64789  
13574       Northern Metropolitan         2474.0   -37.67917, 144.8939  
13575  South-Eastern Metropolitan         7392.0  -37.90562, 145.16761  
13576        Western Metropolitan         6380.0  -37.85927, 144.87904  
13577        Western Metropolitan         6380.0  -37.85274, 144.88738  
13578        Western Metropolitan         6380.0  -37.85908, 144.89299  
13579        Western Metropolitan         6543.0  -37.81188, 144.88449  

[7 rows x 23 columns]

## Размерность таблицы
Далее хотелось бы узнать размер таблицы — количество строк и количество столбцов. Это можно сделать с помощью атрибута shape, который возвращает кортеж с количеством строк и столбцов:

In [32]:

 melb_data.shape
# (13580, 23)

(13580, 23)

In [33]:
melb_data.shape[0]

13580

## Получение информации о столбцах
Для того чтобы получить более детальную информацию о столбцах таблицы, можно использовать метод DataFrame info():
Данный метод выводит:

информацию об индексах;
информацию об общем количестве столбцов;
таблицу, в которой содержится информация об именах столбцов (Column), количестве непустых значений (Non-Null Count) в каждом столбце и типе данных столбца (Dtype), количестве столбцов, в которых используется определённый тип данных;
количество оперативной памяти в мегабайтах, которое тратится на хранение данных.

как работать с пропусками https://pythonru.com/biblioteki/not-a-number-vse-o-nan-pd-5

In [34]:
melb_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  float64
 12  Bathroom       13580 non-null  float64
 13  Car            13580 non-null  float64
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  float64
 17  CouncilArea    12211 non-null  object 
 18  Lattit

## Изменение типа данных в столбце
можно воспользоваться методом astype(), который позволяет преобразовать тип данных столбца:
В данном коде мы при помощи метода astype() последовательно переопределяем столбцы на них же самих, только с изменённым типом данных: int64 — целочисленное число размером 64 бита.
Примечание. При работе с очень большими объёмами данных такая перекодировка может сыграть большую роль, позволяя сэкономить объём занимаемой таблицей памяти.

подробнее о типах данных в Pandas: https://dfedorov.spb.ru/pandas/%D0%9E%D0%B1%D0%B7%D0%BE%D1%80%20%D1%82%D0%B8%D0%BF%D0%BE%D0%B2%20%D0%B4%D0%B0%D0%BD%D0%BD%D1%8B%D1%85%20pandas.html

In [35]:
melb_data['Car'] = melb_data['Car'].astype('int64')
melb_data['Bedroom'] = melb_data['Bedroom'].astype('int64')
melb_data['Bathroom'] = melb_data['Bathroom'].astype('int64')
melb_data['Propertycount'] = melb_data['Propertycount'].astype('int64')
melb_data['YearBuilt'] = melb_data['YearBuilt'].astype('int64')
melb_data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  int64  
 12  Bathroom       13580 non-null  int64  
 13  Car            13580 non-null  int64  
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  int64  
 17  CouncilArea    12211 non-null  object 
 18  Lattit

## Получение описательной статистики

→ Часто при работе с таблицей нужно быстро посмотреть на основные статистические свойства её столбцов. Для этого можно воспользоваться методом DataFrame describe().

По умолчанию метод работает с числовыми (int64 и float64) столбцами и показывает число непустых значений (count), среднее (mean), стандартное отклонение (std), минимальное значение (min),  квартили уровней 0.25, 0.5 (медиана) и 0.75 (25%, 50%, 75%) и максимальное значение (max) для каждого столбца исходной таблицы.
Мы ещё будем подробнее говорить о статистических параметрах в разделе по математике, но советуем вам ознакомиться с данными терминами:

Стандартное отклонение: https://ru.wikipedia.org/wiki/%D0%A1%D1%80%D0%B5%D0%B4%D0%BD%D0%B5%D0%BA%D0%B2%D0%B0%D0%B4%D1%80%D0%B0%D1%82%D0%B8%D1%87%D0%B5%D1%81%D0%BA%D0%BE%D0%B5_%D0%BE%D1%82%D0%BA%D0%BB%D0%BE%D0%BD%D0%B5%D0%BD%D0%B8%D0%B5

Квартиль: https://rus.cubecartskins.net/quartile

In [36]:
# выведем на экран значение статистических параметров только для столбцов Distance (расстояние от объекта недвижимости до центра Мельбурна),
#  BuildingArea (площадь здания) и Price (цена объекта):

melb_data.describe().loc[:, ['Distance', 'BuildingArea' , 'Price']]

Distance  BuildingArea         Price
count  13580.000000  13580.000000  1.358000e+04
mean      10.137776    139.633972  1.075684e+06
std        5.868725    392.217403  6.393107e+05
min        0.000000      0.000000  8.500000e+04
25%        6.100000    122.000000  6.500000e+05
50%        9.200000    126.000000  9.030000e+05
75%       13.000000    129.940000  1.330000e+06
max       48.100000  44515.000000  9.000000e+06

In [37]:
9.030000e+05 / 1.075684e+06 

0.8394658654400363

Столбец Distance (расстояние до центра города)
В нашем наборе данных представлены объекты, удалённые от центра города на расстояние до 48 километров.

Столбец BuildingArea (площадь здания)

В нашей таблице представлены участки, на которых отсутствует само здание (его площадь равна 0), то есть владельцы продавали участок земли без строений на нём.
Большая часть объектов (75%) имеет площадь здания до 130 квадратных метров, однако в нашей таблице есть настоящий дворец (судя по его размерам) с площадью в 44 515 квадратных метров.

Столбец Price (цена)

Диапазон цен на недвижимость в Мельбурне варьируется от 85 тысяч до 9 миллионов австралийских долларов.


На самом деле метод describe() можно применять не только к числовым признакам. С помощью параметра include можно указать тип данных, для которого нужно вывести описательную информацию.

Например, для типа данных object метод describe() возвращает DataFrame, в котором указаны:

количество непустых строк (count);
количество уникальных значений (unique);
самое частое значение — мода —  (top);
частота — объём использования — этого значения (freq) для каждого столбца типа object исходной таблицы.

In [38]:
melb_data.describe(include=['object'])

Suburb           Address   Type Method SellerG        Date  \
count       13580             13580  13580  13580   13580       13580   
unique        314             13378      3      5     268          58   
top     Reservoir  36 Aberfeldie St      h      S  Nelson  27/05/2017   
freq          359                 3   9449   9022    1565         473   

       CouncilArea             Regionname         Coordinates  
count        12211                  13580               13580  
unique          33                      8               13097  
top       Moreland  Southern Metropolitan  -37.8361, 144.9966  
freq          1163                   4695                  12

Также приведём несколько выводов, которые можно сделать из полученной таблицы:

Столбец Suburb (пригород)

Наибольшее количество проданных объектов (359) находилось в пригороде Reservoir.

Столбец Type (тип здания)

Самый популярный тип дома — h (дом, вилла, коттедж).

Столбец SellerG (риелтор)

В наших данных нам известно о 268 различных риэлторских компаниях, однако самой главной «акулой» в этом бизнесе является компания Nelson — они продали 1 565 различных домов.

толбец Date (дата продажи)

В нашей таблице содержится информация за 58 дней, при этом наибольшее число продаж (473) пришлось на 27 мая 2017 года.

## Получние частоты уникальных значений в столбце

Для того чтобы определить, сколько раз в столбце повторяется каждый из вариантов значений (т.е. найти частоту для каждого уникального знания), используется метод value_counts().
Данный метод возвращает объект Series, в котором в качестве индексов выступают уникальные категории столбца, а значениями — соответствующая им частота.

In [39]:
# Рассмотрим работу value_counts() на примере столбца с названиями районов:
melb_data['Regionname'].value_counts()

Southern Metropolitan         4695
Northern Metropolitan         3890
Western Metropolitan          2948
Eastern Metropolitan          1471
South-Eastern Metropolitan     450
Eastern Victoria                53
Northern Victoria               41
Western Victoria                32
Name: Regionname, dtype: int64

normalize

Чтобы сделать вывод более интерпретируемым и понятным, можно воспользоваться параметром normalize. При установке значения этого параметра на True результат будет представляться в виде доли (относительной частоты):

In [40]:

melb_data['Regionname'].value_counts(normalize=True)

Southern Metropolitan         0.345729
Northern Metropolitan         0.286451
Western Metropolitan          0.217084
Eastern Metropolitan          0.108321
South-Eastern Metropolitan    0.033137
Eastern Victoria              0.003903
Northern Victoria             0.003019
Western Victoria              0.002356
Name: Regionname, dtype: float64

In [41]:
melb_data['Distance'].value_counts()

11.2    739
9.2     367
13.9    324
7.8     306
4.6     263
       ... 
37.5      1
29.3      1
47.4      1
39.8      1
48.1      1
Name: Distance, Length: 202, dtype: int64

In [42]:
melb_data['Type'].value_counts(normalize=True)

h    0.695803
u    0.222165
t    0.082032
Name: Type, dtype: float64

# Статистически методы

## Агрегирующие методы

Агрегирующим в Pandas называется метод, который для каждого столбца возвращает только одно значение — показатель (например, вычисление медианы, максимума, среднего и так далее).

Ниже приведена таблица основных агрегирующих методов:

МЕТОД	СТАТИСТИЧЕСКИЙ ПАРАМЕТР
.count()	Количество непустых значений
.mean()	Среднее значение
.min()	Минимальное значение
.max()	Максимальное значение
.deviance()	Дисперсия
.std()	Стандартное отклонение
.sum()	Сумма
.quantile(x)	Квантиль уровня x
.nunique()	Число уникальных значений
.median() - Медиана

Если один из этих методов применить ко всему DataFrame, то в результате его работы будет получен объект типа Series, в котором в качестве индексов будут выступать наименования столбцов, а в качестве значений — статистический показатель. В случае применения метода к отдельному столбцу результатом вычислений станет число.

В каждый метод можно передать некоторые параметры, среди которых:

axis  — определяет, подсчитывать параметр по строкам или по столбцам;
numeric_only — определяет, вычислять параметры только по числовым столбцам/строкам или нет (True/False).


In [43]:
# Вычислим среднюю цену на объекты недвижимости:

print(melb_data['Price'].mean())
# 1075684.079455081

1075684.079455081


In [44]:
# Найдём максимальное количество парковочных мест:
print(melb_data['Car'].max())
# 10

10


In [45]:
# А теперь представим, что риэлторская ставка для всех компаний за продажу недвижимости составляет 12%. Найдём общую прибыльность
#  риэлторского бизнеса в Мельбурне. Результат округлим до сотых:
rate = 0.12
income = melb_data['Price'].sum() * rate
print('Total income of real estate agencies:', round(income, 2))
# Total income of real estate agencies: 1752934775.88

Total income of real estate agencies: 1752934775.88


In [46]:
# Найдём, насколько медианная площадь территории отличается от её среднего значения. Вычислим модуль разницы между медианой и средним 
# и разделим результат на среднее, чтобы получить отклонение в долях:
landsize_median = melb_data['Landsize'].median() 
landsize_mean =  melb_data['Landsize'].mean()
print(abs(landsize_median - landsize_mean)/landsize_mean)
# 0.21205713983546193

0.21205713983546193


## Модальное значение
 Отдельный интерес представляет статический показатель моды — самого распространённого значения в столбце. Он вычисляется с помощью метода mode().
 Модальных значений может быть несколько, то есть несколько значений могут встречаться одинаковое количество раз. Поэтому метод mode(), в отличие от агрегирующих методов, возвращает не одно число, а серию.

In [47]:
# Вычислим, какое число комнат чаще всего представлено на рынке недвижимости:
print(melb_data['Rooms'].mode())
# 0    3
# dtype: int64

0    3
Name: Rooms, dtype: int64


In [48]:
# наиболее распространённое название района
melb_data['Regionname'].mode()

0    Southern Metropolitan
Name: Regionname, dtype: object

## Задания

In [49]:
melb_data['Type'].std()

TypeError: could not convert string to float: 'h'

In [ ]:
# Чему равно максимальное количество домов на продажу в районе (Propertycount)?

melb_data['Propertycount'].max()

#melb_data.head()

21650

In [ ]:
# Чему равно стандартное отклонение (разброс) расстояния от центра города до объекта недвижимости?

melb_data['Distance'].std()

5.868724943071711

In [ ]:
# Чему равно отклонение (в процентах) медианного значения площади здания от его среднего значения?

(melb_data['BuildingArea'].quantile(0.5) / melb_data['BuildingArea'].mean())*100

BuildingArea_median = melb_data['BuildingArea'].median() 
BuildingArea_mean =  melb_data['BuildingArea'].mean()
print((abs(BuildingArea_median - BuildingArea_mean)/BuildingArea_mean)*100)
# 0.21205713983546193



9.764079662364534


In [ ]:
#  [1, 2, 4, 2, 3, 2, 1, 5, 6]

numbers = pd.Series(
    data = [1, 2, 4, 2, 3, 2, 1, 5, 6],
    name = 'numbers'
)
#display(numbers)

numbers.mode()

0    2
Name: numbers, dtype: int64

In [ ]:
# Сколько спален чаще всего встречается в домах в Мельбурне?

melb_data['Bedroom'].mode()

0    3
Name: Bedroom, dtype: int64

# Фильтрация данных в DataFrame

Под фильтрацией в DataFrame подразумевается получение новой таблицы путём вырезания строк, не удовлетворяющих поставленному условию. 

Маской называется Series, которая состоит из булевых значений, при этом значения True соответствуют тем индексам, для которых заданное условие выполняется, в противном случае ставится значение False (например, цена > 2 млн).

In [ ]:
# Создадим маску и положим её в переменную с именем mask. Синтаксис очень прост:
mask = melb_data['Price'] > 2000000
display(mask)

0        False
1        False
2        False
3        False
4        False
         ...  
13575    False
13576    False
13577    False
13578     True
13579    False
Name: Price, Length: 13580, dtype: bool

In [ ]:
# Для фильтрации нужно просто подставить переменную mask в индексацию DataFrame. Маска показывает, какие строки нужно оставлять в 
# результирующем наборе, а какие — убирать (выведем первые пять строк отфильтрованной таблицы):

display(melb_data[mask].head())

index       Suburb               Address  Rooms Type      Price Method  \
80     80  Albert Park  112 Beaconsfield Pde      3    h  2850000.0     PI   
85     85  Albert Park     104 Richardson St      4    h  2300000.0      S   
88     88  Albert Park        29 Faussett St      2    h  2120000.0      S   
92     92  Albert Park           2 Dundas Pl      3    h  2615000.0      S   
93     93  Albert Park          23 Finlay St      5    h  2100000.0      S   

     SellerG        Date  Distance  ...  Car  Landsize  BuildingArea  \
80    Buxton   4/03/2017       3.3  ...    0     211.0         198.0   
85  Marshall   7/05/2016       3.3  ...    1     153.0         180.0   
88      Greg  10/09/2016       3.3  ...    1     199.0         107.0   
92    Cayzer  10/12/2016       3.3  ...    1     177.0         181.0   
93      Greg  10/12/2016       3.3  ...    1     237.0         126.0   

    YearBuilt   CouncilArea  Lattitude  Longtitude             Regionname  \
80       1890  Port Phillip   -37.8481    144.9499  Southern Metropolitan   
85       1880  Port Phillip   -37.8447    144.9523  Southern Metropolitan   
88       1900  Port Phillip   -37.8422    144.9554  Southern Metropolitan   
92       1880  Port Phillip   -37.8415    144.9585  Southern Metropolitan   
93       1970  Port Phillip   -37.8436    144.9557  Southern Metropolitan   

    Propertycount         Coordinates  
80           3280  -37.8481, 144.9499  
85           3280  -37.8447, 144.9523  
88           3280  -37.8422, 144.9554  
92           3280  -37.8415, 144.9585  
93           3280  -37.8436, 144.9557  

[5 rows x 23 columns]

In [ ]:
# Также вовсе не обязательно заносить маску в отдельную переменную — можно сразу вставлять условие в операцию индексации DataFrame, например:

melb_data[melb_data['Price'] > 2000000]

index          Suburb               Address  Rooms Type      Price  \
80        80     Albert Park  112 Beaconsfield Pde      3    h  2850000.0   
85        85     Albert Park     104 Richardson St      4    h  2300000.0   
88        88     Albert Park        29 Faussett St      2    h  2120000.0   
92        92     Albert Park           2 Dundas Pl      3    h  2615000.0   
93        93     Albert Park          23 Finlay St      5    h  2100000.0   
...      ...             ...                   ...    ...  ...        ...   
13521  13521  Port Melbourne          44 Garton St      4    t  2455000.0   
13523  13523         Prahran        69 Greville St      4    h  2668000.0   
13553  13553    Surrey Hills          20 Albert Cr      4    h  2720000.0   
13555  13555    Surrey Hills              3 Oak St      4    h  3100000.0   
13578  13578    Williamstown          96 Verdon St      4    h  2500000.0   

      Method   SellerG        Date  Distance  ...  Car  Landsize  \
80        PI    Buxton   4/03/2017       3.3  ...    0     211.0   
85         S  Marshall   7/05/2016       3.3  ...    1     153.0   
88         S      Greg  10/09/2016       3.3  ...    1     199.0   
92         S    Cayzer  10/12/2016       3.3  ...    1     177.0   
93         S      Greg  10/12/2016       3.3  ...    1     237.0   
...      ...       ...         ...       ...  ...  ...       ...   
13521     SP  Marshall  26/08/2017       3.5  ...    2     123.0   
13523      S    Biggin  26/08/2017       4.6  ...    2     383.0   
13553      S       Kay  26/08/2017      10.2  ...    2    1005.0   
13555     VB  Marshall  26/08/2017      10.2  ...    3     832.0   
13578     PI   Sweeney  26/08/2017       6.8  ...    5     866.0   

       BuildingArea  YearBuilt   CouncilArea  Lattitude  Longtitude  \
80            198.0       1890  Port Phillip  -37.84810   144.94990   
85            180.0       1880  Port Phillip  -37.84470   144.95230   
88            107.0       1900  Port Phillip  -37.84220   144.95540   
92            181.0       1880  Port Phillip  -37.84150   144.95850   
93            126.0       1970  Port Phillip  -37.84360   144.95570   
...             ...        ...           ...        ...         ...   
13521           0.0       2010           NaN  -37.83349   144.94840   
13523         126.0       1970           NaN  -37.84879   144.98882   
13553         126.0       1920           NaN  -37.82421   145.10352   
13555         126.0       1970           NaN  -37.83564   145.10919   
13578         157.0       1920           NaN  -37.85908   144.89299   

                  Regionname  Propertycount           Coordinates  
80     Southern Metropolitan           3280    -37.8481, 144.9499  
85     Southern Metropolitan           3280    -37.8447, 144.9523  
88     Southern Metropolitan           3280    -37.8422, 144.9554  
92     Southern Metropolitan           3280    -37.8415, 144.9585  
93     Southern Metropolitan           3280    -37.8436, 144.9557  
...                      ...            ...                   ...  
13521  Southern Metropolitan           8648   -37.83349, 144.9484  
13523  Southern Metropolitan           7717  -37.84879, 144.98882  
13553  Southern Metropolitan           5457  -37.82421, 145.10352  
13555  Southern Metropolitan           5457  -37.83564, 145.10919  
13578   Western Metropolitan           6380  -37.85908, 144.89299  

[1007 rows x 23 columns]

In [ ]:
# Найдём количество зданий с тремя комнатами. Для этого отфильтруем таблицу по условию: обратимся к результирующей таблице 
# по столбцу Rooms и найдём число строк в ней с помощью атрибута shape:

melb_data[melb_data['Rooms'] == 3].shape[0]
# 5881

5881

In [ ]:
# Усложним прошлый пример и найдём число трёхкомнатных домов с ценой менее 300 тысяч:
melb_data[(melb_data['Rooms'] == 3) & (melb_data['Price'] < 300000)].shape[0]
# 3

3

In [ ]:
# Таких зданий оказалось всего три. Немного «ослабим» условие: теперь нас будут интересовать дома с ценой менее 300 тысяч,
#  у которых либо число комнат равно 3 либо площадь домов более 100 квадратных метров:
melb_data[((melb_data['Rooms'] == 3)| (melb_data['BuildingArea'] > 100)) & (melb_data['Price'] < 300000) ].shape[0]

68

In [ ]:
# Фильтрацию часто сочетают со статистическими методами. Давайте найдём максимальное количество комнат в таунхаусах. 
# Так как в результате фильтрации получается DataFrame, то обратимся к нему по столбцу Rooms и найдём максимальное значение:

melb_data[melb_data['Type'] == 't']['Rooms'].max()

5

In [ ]:
# А теперь более сложный трюк: найдём медианную площадь здания у объектов, чья цена выше средней. 
# Для того чтобы оградить наш код от нагромождений, предварительно создадим переменную со средней ценой:

mean_price = melb_data['Price'].mean()
display(mean_price)

melb_data[melb_data['Price'] > mean_price]['BuildingArea'].median()

1075684.079455081

126.0

## Очистка данных

Фильтрация находит применение в очистке данных.

Под очисткой данных понимается удаление из данных аномальных значений (выбросов), пропусков и данных, которые не несут информацию.

## Задания

In [ ]:
# 8.1
#У скольких объектов недвижимости из таблицы melb_data отсутствуют ванные комнаты?
#melb_data[(melb_data['Bathroom'] == NaN)]

melb_data['Bathroom'].value_counts()


1    7512
2    4974
3     917
4     106
0      34
5      28
6       5
7       2
8       2
Name: Bathroom, dtype: int64

In [ ]:
# 8.2 Сколько в таблице melb_data объектов недвижимости, которые были проданы риелторской компанией Nelson и стоимость которых 
# составила больше 3 миллионов?
melb_data[(melb_data['SellerG'] == 'Nelson') & (melb_data['Price'] > 3000000)].shape[0]

5

In [ ]:
# Какова минимальная стоимость участка без здания (площадь здания равна 0) в таблице melb_data?

melb_data[melb_data['BuildingArea'] == 0]['Price'].min()

412500.0

In [ ]:
# Какова средняя цена объектов недвижимости в таблице melb_data с ценой менее одного миллиона, в которых либо количество 
# комнат больше пяти, либо здание моложе 2015 года?
melb_data[(melb_data['Price'] <  1000000)&((melb_data['Rooms'] > 5) | (melb_data['YearBuilt'] > 2015))]['Price'].mean()

769238.6363636364

In [ ]:
# В каком районе Мельбурна чаще всего продаются виллы и коттеджи (тип здания — h) с ценой меньше трёх миллионов?
melb_data[(melb_data['Type'] == 'h') & (melb_data['Price']  < 3000000)]['Regionname'].mode()

0    Northern Metropolitan
Name: Regionname, dtype: object

# Дополнительная литература

Книга по Pandes  https://itfy.org/threads/kniga-o-biblioteke-pandas-pandas-rabota-s-dannymi-abdraxmanov-m-i.1262/
Введение в Pandas на Habr  https://habr.com/ru/post/196980/